File: stubs/csharp/src/BeastMode.RmDdd.Examples/Examples/ECommerce/ProductId.cs
using System;
using System.Collections.Generic;
using BeastMode.RmDdd.Domain;
using BeastMode.RmDdd.Utilities;

namespace BeastMode.RmDdd.Examples.ECommerce
{
    /// <summary>
    /// Value object representing a Product identifier.
    /// </summary>
    public class ProductId : ValueObject
    {
        public Guid Value { get; }
        
        public ProductId(Guid value)
        {
            Value = value;
            ValidateOnConstruction();
        }
        
        public ProductId(string value) : this(Guid.Parse(value))
        {
        }
        
        public static ProductId Generate()
        {
            return new ProductId(Guid.NewGuid());
        }
        
        public override ValidationResult Validate()
        {
            var result = new ValidationResult();
            
            if (Value == Guid.Empty)
            {
                result.AddError("Product ID value cannot be empty");
            }
            
            return result;
        }
        
        protected override IEnumerable<object?> GetEqualityComponents()
        {
            yield return Value;
        }
        
        public override string ToString()
        {
            return $"ProductId({Value})";
        }
        
        // Implicit conversion operators for convenience
        public static implicit operator Guid(ProductId productId) => productId.Value;
        public static implicit operator ProductId(Guid value) => new(value);
    }
}

File: stubs/java/src/main/java/com/beastmode/rmddd/examples/ecommerce/ProductId.java
package com.beastmode.rmddd.examples.ecommerce;

import com.beastmode.rmddd.domain.ValueObject;
import com.beastmode.rmddd.utilities.ValidationResult;

import java.util.Objects;
import java.util.UUID;

/**
 * Value object representing a Product identifier.
 */
public class ProductId extends ValueObject {
    
    private final UUID value;
    
    public ProductId(UUID value) {
        this.value = Objects.requireNonNull(value, "Product ID value cannot be null");
        validateOnConstruction();
    }
    
    public ProductId(String value) {
        this(UUID.fromString(value));
    }
    
    public static ProductId generate() {
        return new ProductId(UUID.randomUUID());
    }
    
    public UUID getValue() {
        return value;
    }
    
    @Override
    public ValidationResult validate() {
        ValidationResult result = new ValidationResult();
        
        if (value == null) {
            result.addError("Product ID value cannot be null");
        }
        
        return result;
    }
    
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        ProductId productId = (ProductId) o;
        return Objects.equals(value, productId.value);
    }
    
    @Override
    public int hashCode() {
        return Objects.hash(value);
    }
    
    @Override
    public String toString() {
        return "ProductId{" + value + '}';
    }
}